tests/gradients/test_vjp.py

```diff
@@ -0,0 +1,457 @@
+        assert vjp is None
+
+    def test_zero_dy(self):
+        """A zero dy vector will return a zero matrix"""
+        dy = np.zeros([2, 2])
+        jac = np.array([[[1.0, 0.1, 0.2], [0.2, 0.6, 0.1]], [[0.4, -0.7, 1.2], [-0.5, -0.6, 0.7]]])
+
+        vjp = qml.gradients.compute_vjp(dy, jac)
+        assert np.all(vjp == np.zeros([3]))
+
+
+class TestVJP:
+    """Tests for the vjp function"""
+
+    def test_no_trainable_parameters(self):
+        """A tape with no trainable parameters will simply return None"""
+
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(0.4, wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0))
+
+        tape.trainable_params = {}
+        dy = np.array([1.0])
+        tapes, fn = qml.gradients.vjp(tape, dy, param_shift)
+
+        assert not tapes
+        assert fn(tapes) is None
+
+    def test_zero_dy(self):
+        """A zero dy vector will return no tapes and a zero matrix"""
+
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(0.4, wires=0)
+            qml.RX(0.6, wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0))
+
+        tape.trainable_params = {0, 1}
+        dy = np.array([0.0])
+        tapes, fn = qml.gradients.vjp(tape, dy, param_shift)
+
+        assert not tapes
+        assert np.all(fn(tapes) == np.zeros([len(tape.trainable_params)]))
+
+    def test_single_expectation_value(self, tol):
+        """Tests correct output shape and evaluation for a tape
+        with a single expval output"""
+        dev = qml.device("default.qubit", wires=2)
+        x = 0.543
+        y = -0.654
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RX(x, wires=[0])
+            qml.RY(y, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0) @ qml.PauliX(1))
+
+        tape.trainable_params = {0, 1}
+        dy = np.array([1.0])
+
+        tapes, fn = qml.gradients.vjp(tape, dy, param_shift)
+        assert len(tapes) == 4
+
+        res = fn(dev.batch_execute(tapes))
+        assert res.shape == (2,)
+
+        expected = np.array([-np.sin(y) * np.sin(x), np.cos(y) * np.cos(x)])
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_multiple_expectation_values(self, tol):
+        """Tests correct output shape and evaluation for a tape
+        with multiple expval outputs"""
+        dev = qml.device("default.qubit", wires=2)
+        x = 0.543
+        y = -0.654
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RX(x, wires=[0])
+            qml.RY(y, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0))
+            qml.expval(qml.PauliX(1))
+
+        tape.trainable_params = {0, 1}
+        dy = np.array([1.0, 2.0])
+
+        tapes, fn = qml.gradients.vjp(tape, dy, param_shift)
+        assert len(tapes) == 4
+
+        res = fn(dev.batch_execute(tapes))
+        assert res.shape == (2,)

```

Comment:
Interesting that these are both independent of the size of the output!

Reply:
How do you mean?

Reply:
I still don't understand the shapes of everything, but I thought the jacobian has a different shape if the tape output is not a scalar, and I'd expect the vjp to have a different output shape to the previous test... but maybe it's just too late to think straight ;)

Reply:
Oh, this is because it is computing the VJP, and the VJP always has shape `(num_params,)`  -- and both tests have two parameters 🙂 

The dimensional analysis looks like this:

```
vjp[num_params] = dy[output_shape] @ jac[output_shape, num_params]
```

where the `output_shape` axes are being contracted